lib/src/rules/prefer_final_parameters.dart

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc =
+    r'Prefer final for parameter declarations if they are not reassigned.';
+
+const _details = r'''
+
+**DO** prefer declaring parameters as final if they are not reassigned in
+the function body.
+
+Declaring parameters as final when possible is a good practice because it helps
+avoid accidental reassignments.
+
+**BAD:**
+```dart
+void badParameter(String label) { // LINT
+  print(label);
+}
+```
+
+**GOOD:**
+```dart
+void goodParameter(final String label) { // OK
+  print(label);
+}
+```
+
+**BAD:**
+```dart
+void badExpression(int value) => print(value); // LINT
+```
+
+**GOOD:**
+```dart
+void goodExpression(final int value) => print(value); // OK
+```
+
+**GOOD:**
+```dart
+void mutableParameter(String label) { // OK
+  print(label);
+  label = 'Hello Linter!';
+  print(label);
+}
+```
+
+''';
+
+class PreferFinalParameters extends LintRule implements NodeLintRule {
+  PreferFinalParameters()
+      : super(
+            name: 'prefer_final_parameters',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  List<String> get incompatibleRules => const ['unnecessary_final'];
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addFunctionDeclaration(this, visitor);
+    registry.addMethodDeclaration(this, visitor);
+    registry.addFunctionExpression(this, visitor);
```

Comment:
Consider alphabetizing these registry calls.

Do we want to enforce `final` for constructor parameters? If so, this will need to add `ConstructorDeclaration` to the list and `_reportApplicableParameters` will need to allow `FieldFormalParameter`s to not have a `final`.

Reply:
I was planning on support constructor parameters, but completely forgot as I was diving in to the implementation, thanks!